scripts/tests/twister/test_handlers.py

```diff
@@ -0,0 +1,815 @@
+            mock.Mock(fixtures=['dummy fixture'], platform='dummy_platform', serial_pty=mock.Mock(), available=1, counter=0)
+        ],
+        3
+    ),
+    ('dummy_platform', 'dummy fixture', [], TwisterException),
+    (
+        'dummy_platform',
+        'dummy fixture',
+        [mock.Mock(fixtures=['dummy fixture'], platform='dummy_platform', serial_pty=mock.Mock(), available=0)],
+        None
+    )
+]
+
+@pytest.mark.parametrize("platform_name, fixture, duts, expected", TESTDATA_5)
+def test_device_device_is_available(mocked_instance, platform_name, fixture, duts, expected):
+    mocked_instance.platform.name = platform_name
+    mocked_instance.testsuite.harness_config = {'fixture': fixture}
+
+    handler = DeviceHandler(mocked_instance, 'build')
+    handler.duts = duts
+
+    if isinstance(expected, int):
+        device = handler.device_is_available(mocked_instance)
+
+        assert device == duts[expected]
+    elif expected is None:
+        device = handler.device_is_available(mocked_instance)
+
+        assert device is None
+    elif isinstance(expected, type):
+        with pytest.raises(expected):
+            device = handler.device_is_available(mocked_instance)
+    else:
+        assert False
+
+
+def test_device_make_device_available(mocked_instance):
+    serial = mock.Mock(name='dummy_serial')
+    duts = [
+        mock.Mock(available=0, serial=serial, serial_pty=None),
+        mock.Mock(available=0, serial=None, serial_pty=serial),
+        mock.Mock(available=0, serial=mock.Mock('another_serial'), serial_pty=None)
+    ]
+
+    handler = DeviceHandler(mocked_instance, 'build')
+    handler.duts = duts
+
+    handler.make_device_available(serial)
+
+    assert len([None for d in handler.duts if d.available == 1]) == 2
+    assert handler.duts[2].available == 0
+
+
+TESTDATA_6 = [
+    (mock.Mock(pid=0, returncode=0), False),
+    (mock.Mock(pid=0, returncode=1), False),
+    (mock.Mock(pid=0, returncode=1), True)
+]
+
+@pytest.mark.parametrize("mock_process, raise_timeout", TESTDATA_6)
+def test_device_run_custom_script(caplog, mock_process, raise_timeout):
+    def raise_timeout_fn(timeout=-1):
+        if raise_timeout and timeout != -1:
+            raise subprocess.TimeoutExpired(None, timeout)
+        else:
+            return mock.Mock(), mock.Mock()
+
+    def assert_popen(command, stdout=None, stderr=None, cwd=None, env=None, shell=None):
+        return mock.Mock(
+            __enter__=mock.Mock(return_value=mock_process),
+            __exit__=mock.Mock(return_value=None)
+        )
+
+    mock_process.communicate = raise_timeout_fn
+
+    script = mock.Mock()
+    timeout = 60
+
+    with mock.patch("subprocess.Popen", mock.Mock(side_effect=assert_popen)):
+        DeviceHandler.run_custom_script(script, timeout)
+
+    if raise_timeout:
+        assert 'timeout' in caplog.text.lower() or 'timed out' in caplog.text.lower()
```

Comment:
could you please check that `communicate()` was called twice in this case, after `kill()` ?